### PR TITLE
V3.0.3

### LINKED ISSUE #583: DataJsonEncoder. Encode/Decode data to/from Json (similar to DataSerializer)
Format:
`{ type: [type descriptor], value: [data] }`
where:
- `type descriptor` is an arbitrary type (e.g. `Coll[(Coll[Byte], Long)]`);
- `data` in human readable form.

Once you read the type, you know how to interpret the value. 
Examples:
```
{ type: "(Int, Int)"; value: { _1: 10; _2: 20 }}
{ type: "Coll[Byte]"; value: [1, 2, 3] }
{ type: "Coll[(Int, Int)]"; value: { _1: [1, 2, 3]; _2: [10, 20, 30] }}
{ type: "(Coll[(Int, Coll[Byte])], Long)"; 
  value: {
     _1: { _1: [1, 2, 3]; _2: [ [1, 1, 1], [2, 2], [3], [] ] } 
     _2: 100
  }
}
```

### PR DESCRIPTION
**Pull request by kushti**
Includes #587 (close #583 ) only.

### CODE REVIEW

File: src/main/scala/sigmastate/serialization/DataJsonEncoder.scala [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+    val encodedType = tpe.toTermString
+    val encodedData = encodeData(v, tpe)
+    Json.obj(
+      "type" -> Json.fromString(encodedType),
+      "value" -> encodedData,
+    )
+  }
+
+  private def encodeBytes: Encoder[Array[Byte]] = (bytes: Array[Byte]) => {
+    java.util.Base64.getEncoder.encodeToString(bytes).asJson
```
Comment (greenhat) [inline src/main/scala/sigmastate/serialization/DataJsonEncoder.scala]: @gostkin Why Base64 here? Ergo JSON API uses Base16. 
See - http://github.com/ScorexFoundation/sigmastate-interpreter/blob/cf570014ba16bad555a73d71d539568a4cf341a1/src/main/scala/org/ergoplatform/JsonCodecs.scala#L58-L58
Same for `decodeBytes`.

File: src/test/scala/sigmastate/serialization/DataJsonEncoderSpecification.scala [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+import sigmastate.SCollection.SByteArray
+import sigmastate.SType.AnyOps
+import sigmastate.Values.SigmaBoolean
+import sigmastate._
+import sigmastate.eval.Extensions._
+import sigmastate.eval.{Evaluation, _}
+import sigmastate.interpreter.CryptoConstants.EcPointType
+import special.sigma.{AvlTree, Box}
+
+class DataJsonEncoderSpecification extends SerializationSpecification {
```
Comment (greenhat) [inline src/test/scala/sigmastate/serialization/DataJsonEncoderSpecification.scala]: @gostkin A negative test for tuples with more than 2 items would be nice (check out http://github.com/ScorexFoundation/sigmastate-interpreter/blob/cf570014ba16bad555a73d71d539568a4cf341a1/src/test/scala/sigmastate/serialization/ConcreteCollectionSerializerSpecification.scala#L58-L58 for example)

File: src/main/scala/sigmastate/serialization/DataJsonEncoder.scala [hunk 3]
```diff
@@ -32,7 +32,7 @@ object DataJsonEncoder {
   }
 
   private def encodeBytes: Encoder[Array[Byte]] = (bytes: Array[Byte]) => {
-    java.util.Base64.getEncoder.encodeToString(bytes).asJson
+    scorex.util.encode.Base16.encode(bytes).asJson
```
[the remark this answers is not in the document]
Comment (gostkin) [reply src/main/scala/sigmastate/serialization/DataJsonEncoder.scala]: fixed

[the remark this answers is not in the document]
Comment (gostkin) [reply src/main/scala/sigmastate/serialization/DataJsonEncoder.scala]: fixed

**Pull request merged**